### PR TITLE
Change ui test that are run-pass and that do not test the compiler to library tests

### PR DESCRIPTION
**Pull request by CDirkx**
Part of #76268, these are some of the relevant ui tests I found that can be replaced by library tests.

Note: this PR just moves the tests, I have not checked for any overlap between these tests and existing library tests. The only test I changed is `env_home_dir`, where I added code to restore the old home dir after testing.

All moved tests:

| ui test | library test file | test |
| --- | --- | --- |
| `const\ascii_ctype.rs` | `core\tests\ascii.rs` | `ascii_ctype_const` |
| `const\const-str-ptr.rs` | `alloc\tests\str.rs` | `const_str_ptr` |
| `assert-eq-trailing-comma.rs` | `core\tests\macros.rs` | `assert_eq_trailing_comma` |
| `assert-escape.rs` | `core\tests\macros.rs` | `assert_escape` |
| `assert-ne-trailing-comma.rs` | `core\tests\macros.rs` | `assert_ne_trailing_comma` |
| `atomic-access-bool.rs` | `core\tests\atomic.rs` | `atomic_access_bool` |
| `atomic-alignment.rs` | `core\tests\atomic.rs` | `atomic_alignment` |
| `atomic-compare_exchange.rs` | `core\tests\atomic.rs` | `atomic_compare_exchange` |
| ~~`atomic-print.rs`~~ | ~~`std\tests\process.rs`~~ | ~~`atomic_print`~~ |
| `bool.rs` | `core\tests\bool.rs` | `test_bool` |
| `bool_not.rs` | `core\tests\bool.rs` | `test_bool_not` |
| `char_unicode.rs` | `core\tests\unicode.rs` | `version` |
| `cmp-default.rs` | `core\tests\cmp.rs` | `cmp_default` |
| `deref-mut-on-ref.rs` | `core\tests\ops.rs` | `deref_mut_on_ref` |
| `deref-on-ref.rs` | `core\tests\ops.rs` | `deref_on_ref` |
| `env-home-dir.rs` | `std\tests\env.rs` | `env_home_dir` |
| ~~`env-vars.rs`~~ | ~~`std\tests\env.rs`~~ | ~~`env_vars`~~ |
| `extend-for-unit.rs` | `core\tests\iter.rs` | `extend_for_unit` |
| `offset_from.rs` | `core\tests\ptr.rs` | `offset_from` |
| `option-ext.rs` | `core\tests\option.rs` | `option_ext` |
| `result-opt-conversions.rs` | `core\tests\result.rs` | `result_opt_conversions` |
| `sleep.rs` | `std\tests\thread.rs` | `sleep` |
| ~~`try-wait.rs`~~ | ~~`std\tests\process.rs`~~ | ~~`try_wait`~~ |
| `utf8.rs` | `alloc\tests\str.rs` | `utf8` |
| `utf8_chars.rs` | `alloc\tests\str.rs` | `utf8_chars` |
| `wrapping-int-api.rs` | `core\tests\num\wrapping.rs` | `wrapping_int_api` |

### CODE REVIEW

**Comment by CDirkx** [comment]
`env-vars.rs` has been excluded from this move, as when run as a test the other tests could interfere with it.

**Review by dtolnay — APPROVED** [review]
Thank you, this is great.

**Comment by dtolnay** [comment]
@bors r+

Note: not reviewed with full scrutiny, just confirmed that the PR only touches `cfg(test)` code and the +/- line count is sane.

**Comment by CDirkx** [comment]
Excluding `atomic-print.rs` and `try-wait.rs` as well, upon closer inspection these tests were not working as intended. 

These tests communicate with another process spawned using `std::process::Command::new(std::env::current_exe())` and expected specific values for certain `std::env::args()`. However when running as a `#[test]` vs as a ui test the supplied arguments are different. The code could be changed to work as a library test, but I think they are fine staying as ui tests.

**Comment by CDirkx** [comment]
My platform doesn't have 128bit atomics, so I missed a required unstable feature during testing. I double checked all the other tests if I was missing anything else, hopefully this will make everything work on all platforms.

**Comment by jyn514** [comment]
> Note: not reviewed with full scrutiny, just confirmed that the PR only touches cfg(test) code and the +/- line count is sane.

FYI, https://github.com/rust-lang/rust/issues/76268#mentoring-instructions-for-reviewers should make this a lot easier to review without having to read hundreds of lines in detail.

**Comment by CDirkx** [comment]
@rustbot modify labels: +S-waiting-on-review -S-waiting-on-author
Rebased and merged again 😅

File: library/std/tests/env.rs [hunk 1]
```diff
@@ -1,5 +1,6 @@
 use std::env::*;
 use std::ffi::{OsStr, OsString};
+use std::path::PathBuf;
```
Comment (jethrogb) [inline library/std/tests/env.rs:3]: Unused import on cfg(target_env = "sgx")

File: library/alloc/tests/str.rs [hunk 2]
```diff
@@ -1971,3 +1971,97 @@ fn test_str_escapes() {
...
+
+#[test]
+fn const_str_ptr() {
+    const A: [u8; 2] = ['h' as u8, 'i' as u8];
+    const B: &'static [u8; 2] = &A;
+    const C: *const u8 = B as *const u8;
+
+    unsafe {
+        let foo = &A as *const u8;
+        assert_eq!(foo, C);
```
Comment (RalfJung) [inline library/alloc/tests/str.rs:1983]: I am not sure if this equality is truly guaranteed... at least it currently does not hold in Miri. This is probably related to https://github.com/rust-lang/miri/issues/131.

**Pull request merged**